final/model/model_evaluation.py

```diff
@@ -0,0 +1,135 @@
+    the median over the latency values of
+    each prediction.
+    IN:
+        - the path to the tflite model
+        - the path to the original dataset
+    OUT:
+        - preprocess latency
+        - prediction latency
+        - total latency
+        - macro average categorical accuracy
+        - macro average recall
+        - macro average precision
+    """
+    dataset = Dataset(
+        ds_path=args.dataset,
+        preprocess=args.preprocess,
+        batch_size=args.batch_size,
+        sample_duration=30,
+        frame_length_in_s=args.frame_length_in_s,
+        frame_step_in_s=args.frame_step_in_s,
+        random_crop_length_in_s=args.random_crop_length_in_s,
+        num_mel_bins=args.num_mel_bins,
+        lower_frequency=args.lower_frequency,
+        upper_frequency=args.upper_frequency,
+        num_coefficients=args.num_coefficients
+    )
+    latency_preprocess = np.empty(
+        shape=(len(dataset.test_files_ds),),
+        dtype=float
+        )
+    latency_inference = np.empty(
+        shape=(len(dataset.test_files_ds),),
+        dtype=float
+    )
+    model = TFLiteModel(model_path=tflite_model_path)
+    confusion_matrix = np.zeros(shape=(4, 4))
+    print('Starting model evaluation...')
+    for i, filename in tqdm(enumerate(dataset.test_files_ds)):
+        full_audio, label = dataset.get_audio_and_label(filename)
+        window_start = time()
+        audio_segments = sliding_window(
+            signal=full_audio.numpy(),
+            rate=dataset.SAMPLING_RATE,
+            window_size_in_s=args.segment_size_in_s,
+            window_stride_in_s=args.segment_stride_in_s
+            ) if use_sliding_window else [full_audio]
+        preprocess_time = time() - window_start if use_sliding_window else 0
+        inference_time = 0
+        prediction = np.zeros_like(label.numpy())
+        for audio in audio_segments:
+            start_preprocess = time()
+            sample, _ = dataset.preprocess(audio, label)
+            end_preprocess = time()
+            prediction += model.predict(sample)[0]
+            end_inference = time()
+            preprocess_time += end_preprocess - start_preprocess
+            inference_time += end_inference - end_preprocess
+        latency_preprocess[i] = preprocess_time
+        latency_inference[i] = inference_time
+        prediction = np.argmax(prediction)
+        label = np.argmax(label)
+        confusion_matrix[prediction, label] += 1
+    latency_total = latency_preprocess + latency_inference
+    print('Done')
+    latency_indexes = np.argsort(latency_total)
+    if len(latency_indexes) % 2 == 0:
+        median_indexes = latency_indexes[
+            int(len(latency_indexes)/2)-1 : int(len(latency_indexes)/2)+1
+            ]
+    else:
+        median_indexes = latency_indexes[int(len(latency_indexes)/2)]
+    median_latency = np.mean(latency_total[median_indexes])
+    median_latency_preprocess = np.mean(latency_preprocess[median_indexes])
```

Comment:
medians for latency and inference are calculated using indexes that sort total latency, not the individual arrays